scripts/gha/test_simulator.py

```diff
@@ -254,6 +311,135 @@ def _read_file(path):
   return test_result
 
 
+def _check_java_version():
+  command = "java -version 2>&1 | awk -F[\\\"_] 'NR==1{print $2}'"
+  logging.info("Get java version: %s", command)
+  result = subprocess.Popen(command, universal_newlines=True, shell=True, stdout=subprocess.PIPE)
+  java_version = result.stdout.read().strip()
+  logging.info("Java version: %s", java_version)
+  return "1.8" in java_version
+
+
+def _setup_android(platforms_version, build_tools_version, sdk_id):
+  android_home = os.environ["ANDROID_HOME"]
+  pathlist = [os.path.join(android_home, "emulator"), 
+    os.path.join(android_home, "tools"), 
+    os.path.join(android_home, "tools", "bin"), 
+    os.path.join(android_home, "platform-tools"), 
+    os.path.join(android_home, "build-tools", build_tools_version)]
+  os.environ["PATH"] += os.pathsep + os.pathsep.join(pathlist)
+  
+  args = ["sdkmanager", 
+    "emulator", "platform-tools", 
+    "platforms;%s" % platforms_version, 
+    "build-tools;%s" % build_tools_version]
+  logging.info("Install packages: %s", " ".join(args))
+  subprocess.run(args=args, check=True)
+
+  args = ["sdkmanager", sdk_id]
+  logging.info("Download an emulator: %s", " ".join(args))
+  subprocess.run(args=args, check=True)
+
+  args = ["sdkmanager", "--update"]
+  logging.info("Update all installed packages: %s", " ".join(args))
+  subprocess.run(args=args, check=True)
+
+
+def _create_and_boot_emulator(sdk_id):
+  args = ["avdmanager", "-s", 
+    "create", "avd", 
+    "-n", "test_emulator", 
+    "-k", sdk_id, 
+    "-f"]
+  logging.info("Create an emulator: %s", args)
+  p_no = subprocess.Popen(["echo", "no"], stdout=subprocess.PIPE)
+  proc = subprocess.Popen(args, stdin=p_no.stdout, stdout=subprocess.PIPE)
+  p_no.stdout.close()
+  proc.communicate()
+
+  args = ["adb", "start-server"]
+  logging.info("Start adb server: %s", " ".join(args))
+  subprocess.run(args=args, check=True)
+
+  command = "adb devices | grep emulator | cut -f1 | while read line; do adb -s $line emu kill; done"
+  logging.info("Kill all running emulator: %s", command)
+  subprocess.Popen(command, universal_newlines=True, shell=True, stdout=subprocess.PIPE)
+
+  if not FLAGS.ci: 
+    command = "$ANDROID_HOME/emulator/emulator -avd test_emulator &"
+  else:
+    command = "$ANDROID_HOME/emulator/emulator -avd test_emulator -no-window -no-audio -no-boot-anim -gpu swiftshader_indirect &"
+  logging.info("Boot test emulator: %s", command)
+  subprocess.Popen(command, universal_newlines=True, shell=True, stdout=subprocess.PIPE)
```

Comment:
nit: for consistency, we could use `subprocess.run` whenever possible.

Reply:
`subprocess.run` doesn’t work with a background process.

Reply:
Ah! Makes sense.